src/Microsoft.JSInterop/DotNetDispatcher.cs

```diff
@@ -0,0 +1,181 @@
+
+            try
+            {
+                return methodInfo.Invoke(null, suppliedArgs);
+            }
+            catch (Exception ex)
+            {
+                throw UnwrapException(ex);
+            }
+        }
+
+        /// <summary>
+        /// Receives notification that a call from .NET to JS has finished, marking the
+        /// associated <see cref="Task"/> as completed.
+        /// </summary>
+        /// <param name="asyncHandle">The identifier for the function invocation.</param>
+        /// <param name="succeeded">A flag to indicate whether the invocation succeeded.</param>
+        /// <param name="resultOrException">If <paramref name="succeeded"/> is <c>true</c>, specifies the invocation result. If <paramref name="succeeded"/> is <c>false</c>, gives the <see cref="Exception"/> corresponding to the invocation failure.</param>
+        [JSInvokable(nameof(DotNetDispatcher) + "." + nameof(EndInvoke))]
+        public static void EndInvoke(long asyncHandle, bool succeeded, object resultOrException)
+            => ((JSRuntimeBase)JSRuntime.Current).EndInvokeJS(asyncHandle, succeeded, resultOrException);
+
+        private static (MethodInfo, Type[]) GetCachedMethodInfo(string assemblyName, string methodIdentifier)
+        {
+            if (string.IsNullOrWhiteSpace(assemblyName))
+            {
+                throw new ArgumentException("Cannot be null, empty, or whitespace.", nameof(assemblyName));
+            }
+
+            if (string.IsNullOrWhiteSpace(methodIdentifier))
+            {
+                throw new ArgumentException("Cannot be null, empty, or whitespace.", nameof(methodIdentifier));
+            }
+
+            var assemblyMethods = _cachedMethodsByAssembly.GetOrAdd(assemblyName, ScanAssemblyForCallableMethods);
+            if (assemblyMethods.TryGetValue(methodIdentifier, out var result))
+            {
+                return result;
+            }
+            else
+            {
+                throw new ArgumentException($"The assembly '{assemblyName}' does not contain a public method with [{nameof(JSInvokableAttribute)}(\"{methodIdentifier}\")].");
+            }
+        }
+
+        private static IReadOnlyDictionary<string, (MethodInfo, Type[])> ScanAssemblyForCallableMethods(string assemblyName)
+        {
+            // TODO: Consider looking first for assembly-level attributes (i.e., if there are any,
+            // only use those) to avoid scanning, especially for framework assemblies.
+            return GetRequiredLoadedAssembly(assemblyName)
+                .GetExportedTypes()
+                .SelectMany(type => type.GetMethods())
+                .Where(method => method.IsDefined(typeof(JSInvokableAttribute), inherit: false))
+                .ToDictionary(
+                    method => method.GetCustomAttribute<JSInvokableAttribute>(false).Identifier,
+                    method => (method, method.GetParameters().Select(p => p.ParameterType).ToArray())
```

Comment:
I'm not too worried about that, as it will already throw with a clear message when you try to actually invoke it (e.g., "not a closed generic type"). If we try to anticipate all the cases here there might be a lot to consider.